### PR TITLE
feature(NodeAffinity): return Skip in PreScore when nothing to do in Score

### LINKED ISSUE #115745: [Umbrella] Skip Score plugins when coupled PreScore plugin returns Skip status
### What would you like to be added?

It's much similar to https://github.com/kubernetes/kubernetes/issues/114399.
In https://github.com/kubernetes/kubernetes/issues/114827, we support Skip in PreScore/Score as well. (Thanks @kidddddddddddddddddddddd)
So, the plugin can return Skip in PreScore when it doesn't do anything in Score in that scheduling cycle.

For example, if the pod doesn't have any preferred inter-pod affinity, the inter-pod affinity plugin PreScore can return Skip, and the scheduler won't run the inter-pod affinity plugin Score.

Let's change our in-tree Score plugins to return Skip.

Here, I just listed all score plugins, **not all of them may not need to return Skip in PreScore though**. Assignee, please make sure your PreScore/Score plugin really needs to return Skip or not.

Please feel free to assign what you'd like to work on. And **please comment in this issue before starting to work on so that we can avoid many people from working on the same thing!**

- [x] ImageLocality @tangwz  no need.
- [x] InterPodAffinity @utam0k https://github.com/kubernetes/kubernetes/pull/117794
- [x] NodeAffinity @sanposhiho https://github.com/kubernetes/kubernetes/pull/117024
- [x] Fit @tangwz no need.
- [x] BalancedAllocation @tangwz no need.
- [x] PodTopologySpread @utam0k https://github.com/kubernetes/kubernetes/pull/118608
- [x] SelectorSpread; no need. It'll soon be removed. https://github.com/kubernetes/kubernetes/issues/108444
- [x] TaintToleration @tangwz no need. https://github.com/kubernetes/kubernetes/pull/118324
- [x] VolumeBinding @kidddddddddddddddddddddd https://github.com/kubernetes/kubernetes/pull/115768

### Why is this needed?

The scheduler can know which plugins it really needs/doesn't need to run to get the scheduling result.
It'd contribute to the scheduling performance.

/sig scheduling
/assign

### PR DESCRIPTION
**Pull request by sanposhiho**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR make NodeAffinity PreScore returning Skip when nothing to do in NodeAffinity Score

#### Which issue(s) this PR fixes:

part of #115745

#### Does this PR introduce a user-facing change?

```release-note
Scheduler skips NodeAffinity Score plugin when NodeAffinity Score plugin has nothing to do with a Pod.
You might notice an increase in the metric plugin_execution_duration_seconds for extension_point=score plugin=NodeAffinity, because the plugin will only run when the plugin is relevant
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by AxeZhan** [comment]
> Fixes #115745

Part of #115745?

**Comment by sanposhiho** [comment]
@denkensk @chendave Can you have time to take a look at it?

File: pkg/scheduler/framework/plugins/nodeaffinity/node_affinity_test.go [hunk 1]
```diff
@@ -1021,8 +1023,11 @@ func TestNodeAffinityPriority(t *testing.T) {
 				{ObjectMeta: metav1.ObjectMeta{Name: "node2", Labels: label2}},
 				{ObjectMeta: metav1.ObjectMeta{Name: "node3", Labels: label3}},
 			},
-			expectedList: []framework.NodeScore{{Name: "node1", Score: 0}, {Name: "node2", Score: 0}, {Name: "node3", Score: 0}},
-			runPreScore:  true,
+			// If PreScore is enabled, the scheduler doesn't execute Score actually.
+			// But, in case PreScore is disabled, Score returns 0 scores to all Nodes.
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/nodeaffinity/node_affinity_test.go:1027]: I am not sure why this sentence is relevant to this test?

Comment (chendave) [reply pkg/scheduler/framework/plugins/nodeaffinity/node_affinity_test.go:1027]: echo with the comments here, maybe you should keep the original testcases with `runPreScore` set to `false` and just create another one to validate the `Skip` status.

File: pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go [hunk 2]
```diff
@@ -192,6 +192,10 @@ func (pl *NodeAffinity) PreScore(ctx context.Context, cycleState *framework.Cycl
 	if err != nil {
 		return framework.AsStatus(err)
 	}
```
Comment (chendave) [inline pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go:194]: ideally, we'd better have a testcase to cover this code branch, i.e. the `PreScore` is enabled, but expected an `err` status.

**Review by chendave — COMMENTED** [review]
sorry for missing this one, just some suggestion otherwise LGTM.

File: pkg/scheduler/framework/plugins/nodeaffinity/node_affinity_test.go (hunk 1 above)
Comment (sanposhiho) [reply pkg/scheduler/framework/plugins/nodeaffinity/node_affinity_test.go:1027]: > you should keep the original testcases with runPreScore set to false and just create another one to validate the Skip status.

sg.

File: pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go (hunk 2 above)
Comment (sanposhiho) [reply pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go:194]: done

**Comment by sanposhiho** [comment]
@chendave @ahg-g Sorry, I completely forgot about this PR. I addressed your comments, PTAL 🙏

**Comment by ahg-g** [comment]
> @chendave @ahg-g Sorry, I completely forgot about this PR. I addressed your comments, PTAL 🙏

integration tests are failing

**Comment by sanposhiho** [comment]
OK, there is a bug on the framework side... 😓

**Comment by sanposhiho** [comment]
The failing e2e will be resolved by https://github.com/kubernetes/kubernetes/pull/118297.

**Comment by sanposhiho** [comment]
@ahg-g @chendave PTAL 🙏

**Review by chendave — COMMENTED** [review]
/lgtm

**Comment by sanposhiho** [comment]
Fixed @chendave 's point.

@kubernetes/sig-scheduling-approvers Please anyone take a look for `/approve`

**Comment by alculquicondor** [comment]
```
It may affect some metrics values related to the NodeAffinity Score plugin.
```
what do you mean by this?

**Comment by sanposhiho** [comment]
> what do you mean by this?

In metrics, after this PR is merged, it probably looks like NodeAffinity's Score plugin's latency go **increasing** because we execute NodeAffinity Score plugin only when it calculates something.

**Comment by alculquicondor** [comment]
Ok, that's not very clear in the release notes. Maybe reword like this:

```
You might notice an increase in the metric <name-of-metric> for endpoint(?)=score, because the plugin will only run when the plugin is relevant
```

**Comment by sanposhiho** [comment]
@alculquicondor Updated.
